Update fee and dust handling for zero fee channels

carlaKC opened the pull request:
This PR completes the off-chain handling of V3 channels, as described in #3789.

carlaKC commented on lightning/src/ln/channel.rs:
```diff
@@ -5078,6 +5092,7 @@ where
 	) -> u64 {
 		if funding.get_channel_type().supports_anchor_zero_fee_commitments() {
 			debug_assert_eq!(self.feerate_per_kw, 0);
+			debug_assert!(fee_spike_buffer_htlc.is_none());
```
Wondering whether it's worth having this assertion (which makes it necessary to check channel type before calling fee functions). The alternative would be to just ignore the parameters completely for zero fee channels (even though `Some` `fee_spike_buffer_htlc` doesn't make sense for the type).

elnosh commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -244,6 +246,22 @@ pub(crate) fn htlc_tx_fees_sat(feerate_per_kw: u32, num_accepted_htlcs: usize, n
 }
 
+/// Returns a fee estimate for the commitment transaction depending on channel type.
+pub(super) fn commitment_sat_per_1000_weight_for_type<'a, F: Deref>(
+	fee_estimator: &'a LowerBoundedFeeEstimator<F>, channel_type: &ChannelTypeFeatures,
+) -> u32
```
I don't think this lifetime is necessary.

elnosh commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -829,7 +829,7 @@ pub(crate) fn build_htlc_output(
 		htlc_success_tx_weight(channel_type_features)
 	};
-	let output_value = if channel_type_features.supports_anchors_zero_fee_htlc_tx() && !channel_type_features.supports_anchors_nonzero_fee_htlc_tx() {
+	let output_value = if channel_type_features.supports_anchors_zero_fee_htlc_tx() {
```
This commit seems unnecessary given that in the following one it ends up removing this check.

carlaKC replied:
It's pedantic, but I like to aim for single responsibility commits. If we happened to introduce a bug with this PR, this would come in handy for bisecting and knowing that removing this redundant check _wasn't_ the problem.

elnosh commented on lightning/src/ln/channel.rs:
```diff
@@ -4414,14 +4414,22 @@ where
 			// A `None` `HTLCCandidate` is used as in this case because we're already accounting for
 			// the incoming HTLC as it has been fully committed by both sides.
-			let mut remote_fee_cost_incl_stuck_buffer_msat = self.next_remote_commit_tx_fee_msat(funding, None, Some(()));
+			let fee_spike_buffer_htlc = if funding.get_channel_type().supports_anchor_zero_fee_commitments() {
+				None
+			} else {
+				Some(())
+			};
+
+			let mut remote_fee_cost_incl_stuck_buffer_msat = self.next_remote_commit_tx_fee_msat(
+				funding, None, fee_spike_buffer_htlc,
+			);
```
seems this call to `self.next_remote_commit_tx_fee_msat` could stay as it was? It doesn't affect if `fee_spike_buffer_htlc` is `Some` or `None` since this check:

```
if funding.get_channel_type().supports_anchor_zero_fee_commitments() {
	debug_assert_eq!(context.feerate_per_kw, 0);
	debug_assert!(fee_spike_buffer_htlc.is_none());
	return 0
}
```
in `next_remote_commit_tx_fee_msat` happens beforehand. So it will return 0 regardless of the `fee_spike_buffer_htlc`

carlaKC replied:
IMO the caller shouldn't rely on any "inside information" about how the function works, it should just pass the correct information based on its context, and in this case a zero fee channel doesn't need a fee buffer. 

I was [also on the fence about this](https://github.com/lightningdevkit/rust-lightning/pull/3884/commits/5fe6fa7278572b26c1513bb1358b4f8f5a71d2bc#r2162418557), so happy to change it if there are strong feelings here!

elnosh replied:
> it should just pass the correct information based on its context

Agree. Just mentioned it since it felt redundant to do the same check but no strong feelings. Cool to leave as is

carlaKC commented on lightning/src/ln/channel.rs:
```diff
@@ -5462,20 +5454,7 @@ where
-		// channel with non-zero fees, we could produce a non-standard commitment transaction that
-		// puts us at risk of losing funds. We would expect our peer to reject such a channel
-		// open, but we don't want to rely on their validation.
-		assert!(!next_channel_type.supports_anchor_zero_fee_commitments());
```
Yes I think so - previously the motivation for the assert was that if we hit this we'd create an un-broadcastable zero fee commitment channel (because it has non-zero fees). Now that we're properly setting a zero fee, that risk is gone.

We still don't expect to get the zero fee channel type here though (since it's our "highest" chan type) - perhaps downgrade to a `debug_assert`?

tankyleo replied:
Thanks for clarifying ! I'm good to delete it entirely like you have it here.

elnosh commented on lightning/src/ln/channel.rs:
```diff
@@ -4331,6 +4331,10 @@ where
 		F::Target: FeeEstimator,
 		L::Target: Logger,
 	{
+		if funding.get_channel_type().supports_anchor_zero_fee_commitments() {
+			return false;
+		}
+
```
already mentioned by @tankyleo but perhaps this could be accomplished by skipping it in `update_channel_fee` in the channelmanager?

carlaKC commented on lightning/src/ln/update_fee_tests.rs:
```diff
@@ -1375,3 +1375,59 @@ pub fn do_can_afford_given_trimmed_htlcs(inequality_regions: core::cmp::Ordering
+	};
+	assert_zero_fee();
+
+	// Sender should not queue an update_fee message.
+	nodes[0].node.timer_tick_occurred();
+	let events_0 = nodes[0].node.get_and_clear_pending_msg_events();
+	assert_eq!(events_0.len(), 0);
```
> Looks like we should update the condition to skip a fee update in fn update_channel_fee:

I don't _really_ feel like it's channel manger's responsibility to know that we shouldn't be sending fee updates for zero fee commitments (belongs in channel IMO), but checking "has this fee changed" is fair game so will look at updating 👍

tankyleo commented on lightning/src/ln/channel.rs:
```diff
@@ -7969,6 +7979,12 @@ where
 			return Ok(());
 		}
+
+		if self.funding.get_channel_type().supports_anchor_zero_fee_commitments() {
```
Gave it some thought - I think you are right, the current setup might be best :) Feel free to resolve if you are satisfied too.

carlaKC commented:
[Addressed review](https://github.com/lightningdevkit/rust-lightning/compare/5fe6fa7278572b26c1513bb1358b4f8f5a71d2bc..f76a270ac6d2532c288d5313594e8acbefece3a8): only major change is moving responsibility for not triggering `update_fee` on zero fee channels to `channelmanger` (as there's been no change between new/old fee) and changing `channel` to just assert that zero fee channels don't hit update code.

tankyleo reviewed (commented):
LGTM mod two nits, feel free to squash / ping second reviewer :)

carlaKC commented on lightning/src/ln/chan_utils.rs:
```diff
@@ -233,15 +235,45 @@ pub(crate) fn commit_tx_fee_sat(feerate_per_kw: u32, num_htlcs: usize, channel_t
+}
+
+/// Returns a fee estimate for the commitment transaction depending on channel type.
+pub(super) fn commitment_sat_per_1000_weight_for_type<F: Deref>(
```
Can you think of anything less atrociously long than `holder_preferred_commitment_sat_per_1000_weight_for_type`?
